keps/140-scenario-based-simulation/README.md

```diff
@@ -0,0 +1,397 @@
+
+Now that simulators are no longer just for webUI, we need to think about how we can design scenario-based simulation to be easy to use from other clients as well.
+
+Therefore, this kep proposes to define the scenario **as CRD**. All clients, including web UI, can use the scenario-based simulation feature by creating the Scenario resource.
+
+#### Scenario CRD
+
+The Scenario is a non-namespaced resource. 
+This CRD will be applied to kube-apiserver started in kube-scheduler-simulator.
+
+We may need to change etcd request-size limitation by --max-request-bytes since the scenario resource may be bigger than other normal resources.
+https://etcd.io/docs/v3.4/dev-guide/limit/#request-size-limit
+
+```go
+type Scenario struct {
+  metav1.TypeMeta 
+  metav1.ObjectMeta // non namespaced.
+
+  Spec ScenarioSpec
+  Status ScenarioStatus
+}
+
+type ScenarioSpec struct {
+  // Events field has all operations for a scenario.
+  // Also you can add new events during the scenario is running.
+  //
+  // +patchMergeKey=ID
+  // +patchStrategy=merge
+  Events   []ScenarioEvent
+}
+
+type ScenarioEvent struct {
+  // ID for this event. Normally, the system sets this field for you.
+  ID string
+  // Time indicates the time at which the event occurs.
+  Time FakeTime
+  // Operation describes which operation this event wants to do.
+  // Only "Create", "Patch", "Delete", "Done" are valid operations in ScenarioEvent.
+  Operation OperationType
+
+  // One of the following four fields must be specified.
+  // If more than one is specified or if all are empty, the event is invalid and the scenario will fail.
+
+  // CreateOperation is the operation to create new resource.
+  // When use CreateOperation, Operation should be "Create".
+  CreateOperation *CreateOperation
+  // PatchOperation is the operation to patch a resource.
+  // When use PatchOperation, Operation should be "Patch".
+  PatchOperation *PatchOperation
+  // DeleteOperation indicates the operation to delete a resource.
+  // When use DeleteOperation, Operation should be "Delete".
+  DeleteOperation *DeleteOperation
+  // DoneOperation indicates the operation to mark the scenario as DONE.
+  // When use DoneOperation, Operation should be "Done".
+  DoneOperation   *DoneOperation
+}
+
+// OperationType describes Operation.
+// Please see the following defined OperationType, all operation types not listed below are invalid.
+type OperationType string 
+
+const (
+  CreateOperation         OperationType = "Create"
+  PatchOperation          OperationType = "Patch"
+  DeleteOperation         OperationType = "Delete"
+  DoneOperation           OperationType = "Done"
+  PodScheduledOperation   OperationType = "PodScheduled"
+  PodUnscheduledOperation OperationType = "PodUnscheduled"
+  PodPreemptedOperation   OperationType = "PodPreempted"
+)
+
+type CreateOperation struct {
+  // Object is the Object to be create.
+  Object runtime.Object
```

Comment:
nits: is `unstructured.Unstructured` better because `runtime.Object` is interface?

Reply:
Hmm, but all `Object` here should be serialized into some Kubernetes object known by scheme.